Initial L7 Documentation (#6056)
hashicorp/consul: website/source/docs/connect/discovery-chain.html.md

---
layout: "docs"

> page_title: "Connect - Discovery Chain (beta)"

Reviewer: ```suggestion
page_title: "Connect - L7 Traffic Management (beta)"
```

sidebar_current: "docs-connect-discovery_chain"
description: |-

>   A discovery chain is a series of distinct stages that a service discovery request flows through to discover and configure Connect proxy upstreams of the service type.

Reviewer: ```suggestion
 Layer 7 traffic management allows operators to divide L7 traffic between different subsets of service instances. The discovery chain is a series of distinct stages that a service discovery request flows through to discover subsets of Connect proxy upstream services.
```

---

# Discovery Chain <sup>(beta)</sup>

-> **Note:** This feature is not compatible with the 
[built-in proxy](/docs/connect/proxies/built-in.html)
or [native proxies](/docs/connect/native.html).

A discovery chain is a series of distinct stages that a service discovery

> request flows through to discover and configure Connect proxy upstreams of the

Reviewer: Unless the chain itself configures connect proxy upstreams this English is a bit weird. As I understand it the discovery chain is made of config entries. Config entries configure proxy upstreams.

`service` type.

![diagram of discovery chain stages](/assets/images/discovery-chain-simple.svg)

Each stage of this discovery chain can be dynamically reconfigured via various
[configuration entries](/docs/agent/config_entries.html). When a configuration
entry is missing, that stage will fall back on reasonable default behavior.

## Terminology

### Service Subsets

> There are many ways you may wish to carve up a single datacenter's pool of

Reviewer: This is great. I'd lead with this paragraph as the first one on the page, and explain everything else with the context of enabling these use cases :D

services beyond simply returning all healthy instances for load balancing.
Canary testing, A/B tests, blue/green deploys, and soft multi-tenancy
(prod/qa/staging sharing compute resources) all require some mechanism of
carving out portions of the Consul catalog smaller than the level of a single

> service.

Reviewer: The page might flow well if after this paragraph you talk very briefly about service subsets here and then describe the discovery chain as a way to route services to specific subsets. Then you can go into the services of the discovery chain stages and link people out to the config entry kinds where they can get more info... Hope that makes sense...

Reviewer: So it would be:

1. Use cases
2. Subsets
3. Discovery chain allows you to discover subsets
4. Steps in the discovery chain with links to config entries

Reviewer: Somewhere at the end of this paragraph, I think it would also be good to mention that you need Connect to do this stuff and why. It seems obvious because the docs are in the connect docs, but some people still could wonder.

> A service subset assigns a concrete name to a specific partitioning of the

Reviewer: ```suggestion
A service subset assigns a concrete name to a specific subset
```

Reviewer: Most of the talk about subsets should move down to the resolver page or the resolving portion of this page. We just want to give people a big-picture idea of what they are here

> overall set of available service instances within a datacenter used during

Reviewer: ```suggestion
of discoverable service instances within a datacenter.
```

> service discovery.

Reviewer: ```suggestion
```

> A service subset name is useful only when composed with an actual service name,

Reviewer: It might be good to provide an example of what a subset name looks like here

a specific datacenter, and namespace.

All services have an unnamed default subset that will return all healthy
instances unfiltered.

Subsets are defined in
[`service-resolver`](/docs/agent/config-entries/service-resolver.html)
configuration entries, but are referenced by their names throughout the other
configuration entry kinds.

## Stages

### Routing

-> **Note:** A service must define its protocol to be http-based to configure a router.

A [`service-router`](/docs/agent/config-entries/service-router.html) config
entry kind represents the topmost part of the discovery chain.

You can use this to intercept traffic using layer-7 criteria (such as path
prefixes or http headers) and change behavior such as sending traffic to a
different service or service subset.

> These config entries may only reference `service-splitter` or

Reviewer: Did we talk about what it means to "reference" something?

`service-resolver` entries.

### Splitting

-> **Note:** A service must define its protocol to be http-based to configure a splitter.

A [`service-splitter`](/docs/agent/config-entries/service-splitter.html) config
entry kind represents the next hop of the discovery chain after routing.

If no splitter config is defined for a service it is assumed 100% of traffic
flows to a service with the same name as the chain and discovery continues on
to the resolution stage.

> A splitter config entry allows for a user to choose to split incoming requests

Reviewer: <3 examples

Author: I linked out to the subsection on each config entry's page that already has some examples rather than duplicating.
